docs/en/ingest-management/processors/processor-add_kubernetes_metadata.asciidoc

```diff
@@ -0,0 +1,229 @@
+[[add_kubernetes_metadata-processor]]
+= Add Kubernetes metadata
+
+++++
+<titleabbrev>add_kubernetes_metadata</titleabbrev>
+++++
+
+// REVIEWERS: Is there really a situation where users need to configure this
+// processor?
+// If so, what are the implications on default indexers and matchers? Do they
+// exist, or does the user always need to define them if they are adding this
+// processor?
+
+// Also I think the field names need to be updated here. What should they be?
+
+//TODO: This content needs more work, but it's pretty complicated, so I will wait
+//until I get confirmation that we want to add this to the agent docs.
+
+TIP: Inputs that collect logs and metrics use this processor by default, so you
+do not need to configure it explicitly. 
+
+The `add_kubernetes_metadata` processor annotates each event with relevant
+metadata based on which Kubernetes Pod the event originated from. At startup it
+detects an `in_cluster` environment and caches the Kubernetes-related metadata.
+
+For events to be annotated with Kubernetes-related metadata, the Kubernetes
+configuration must be valid.
+
+Each event is annotated with:
+
+* Pod Name
+* Pod UID
+* Namespace
+* Labels
+
+In addition, the node and namespace metadata are added to the Pod metadata.
+
+The `add_kubernetes_metadata` processor has two basic building blocks:
+
+* Indexers
+* Matchers
+
+Indexers use Pod metadata to create unique identifiers for each one of the Pods.
+These identifiers help to correlate the metadata of the observed Pods with
+actual events. For example, the `ip_port` indexer can take a Kubernetes Pod and
+create identifiers for it based on all its `pod_ip:container_port` combinations.
+
+Matchers use information in events to construct lookup keys that match the
+identifiers created by the indexers. For example, when the `fields` matcher
+takes `["metricset.host"]` as a lookup field, it constructs a lookup key with
+the value of the field `metricset.host`. When one of these lookup keys matches
+with one of the identifiers, the event is enriched with the metadata of the
+identified Pod.
+
+//TODO: Figure out what, if anything, to say about default indexers and matchers.
```

Comment:
Need input from reviewers on this question: Are there default indexers and matchers, or do users need to define them explicitly?

Reply:
https://www.elastic.co/guide/en/beats/filebeat/current/add-kubernetes-metadata.html#kubernetes-indexers-and-matchers and https://www.elastic.co/guide/en/beats/metricbeat/current/add-kubernetes-metadata.html#kubernetes-indexers-and-matchers

Reply:
There is a reference inside these docs. Look for example for `When add_kubernetes_metadata is used with Metricbeat, it uses the`

Reply:
This is confusing indeed,

Agent for metrics : (from metricbeat) it uses the ip_port indexer and the fields matcher with the metricset.host field. 
Agent for logs: (from filebeat) it uses the container indexer and matcher logs_path

The confusing part here is that we have a reference for metricset, but we can leave it as it is for now.

Reply:
OK it sounds like there is no action to take right now. Perhaps we can revisit this content when we work on improvements to the K8s monitoring docs.